src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -0,0 +1,483 @@
+        {
+            var solution = currentProject.Solution;
+            var cacheService = GetCacheService(solution.Workspace);
+            var graph = currentProject.Solution.GetProjectDependencyGraph();
+            var relevantProjectIds = graph.GetProjectsThatThisProjectTransitivelyDependsOn(currentProject.Id)
+                                        .Concat(currentProject.Id);
+
+            using var syntaxDisposer = ArrayBuilder<CacheEntry>.GetInstance(out var syntaxBuilder);
+            using var symbolDisposer = ArrayBuilder<SymbolTreeInfo>.GetInstance(out var symbolBuilder);
+            var peReferences = PooledHashSet<PortableExecutableReference>.GetInstance();
+
+            // We will only create missing indices in the following cases, neither would block completion.
+            // 1. We are asked explicitly to create missing indices (e.g. via expander) 
+            // 2. We are trying to populate the data in background.
+            var shouldCreateIndex = forceIndexCreation || isPrecalculation;
+
+            try
+            {
+                foreach (var projectId in relevantProjectIds)
+                {
+                    var project = solution.GetProject(projectId);
+                    if (project == null || !project.SupportsCompilation)
+                    {
+                        continue;
+                    }
+
+                    // Transitively get all the PE references
+                    peReferences.AddRange(project.MetadataReferences.OfType<PortableExecutableReference>());
+
+                    // By default, don't trigger index creation except for documents in current project.
+                    var loadOnly = !shouldCreateIndex && projectId != currentProject.Id;
+                    var cacheEntry = await GetCacheEntryAsync(project, loadOnly, cacheService, cancellationToken).ConfigureAwait(false);
+
+                    // Don't provide anything if we don't have all the required SyntaxTreeIndex created.
+                    if (cacheEntry == null)
+                    {
+                        return null;
+                    }
+
+                    syntaxBuilder.Add(cacheEntry.Value);
+                }
+
+                foreach (var peReference in peReferences)
+                {
+                    var info = await SymbolTreeInfo.GetInfoForMetadataReferenceAsync(
+                        solution, peReference, loadOnly: !shouldCreateIndex, cancellationToken).ConfigureAwait(false);
+
+                    // Don't provide anything if we don't have all the required SymbolTreeInfo created.
+                    if (info == null)
+                    {
+                        return null;
+                    }
+
+                    if (info.ContainsExtensionMethod)
+                    {
+                        symbolBuilder.Add(info);
+                    }
+                }
+
+                // We are just trying to populate the cache in background, no need to return any results.
```

Comment:
ew.  this is not pretty.  a better way would be to extract out the cache computation logic and then call into that from both locations rather than having the precalculation logic cause it to happen (but hten skip some work) by calling into the actual *query* codepath.

Reply:
fixed